osutil/epoll/epoll.go

```diff
@@ -0,0 +1,151 @@
+// Package epoll contains a thin wrapper around the epoll(7) facility.
+//
+// Using epoll from Go is unusual as the language provides facilities that
+// normally make using it directly pointless. Epoll is strictly required for
+// unusual kernel interfaces that use event notification but don't implement
+// file descriptors that provide usual read/write semantics.
+//
+// It might be possible to remove this code and use internal/poll and
+// syscall.RawConn but the necessary interfaces require golang 1.12 that snapd
+// cannot yet depend on.
+package epoll
+
+import (
+	"runtime"
+	"strings"
+
+	"golang.org/x/sys/unix"
+)
+
+// Readiness is the bit mask of aspects of readiness to monitor with epoll.
+type Readiness int
+
+const (
+	// Readable indicates readiness for reading (EPOLLIN).
+	Readable Readiness = unix.EPOLLIN
+	// Writable indicates readiness for writing (EPOLLOUT).
+	Writable Readiness = unix.EPOLLOUT
+)
+
+// String returns readable representation of the readiness flags.
+func (r Readiness) String() string {
+	frags := make([]string, 0, 2)
+	if r&Readable != 0 {
+		frags = append(frags, "Readable")
+	}
+	if r&Writable != 0 {
+		frags = append(frags, "Writable")
+	}
+	return strings.Join(frags, "|")
+}
+
+// Epoll wraps a file descriptor which can be used for I/O readiness notification.
+type Epoll struct {
+	fd        int
+	sysEvents []unix.EpollEvent
+}
+
+// Open opens an event monitoring descriptor.
+func Open() (*Epoll, error) {
+	fd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, err
+	}
+	e := &Epoll{
+		fd:        fd,
+		sysEvents: make([]unix.EpollEvent, 0, 10),
+	}
+	runtime.SetFinalizer(e, func(e *Epoll) {
+		if e.fd != -1 {
+			e.Close()
+		}
+	})
+	return e, nil
+}
+
+// Close closes the event monitoring descriptor.
+func (e *Epoll) Close() error {
+	runtime.SetFinalizer(e, nil)
+	fd := e.fd
+	e.fd = -1
+	return unix.Close(fd)
+}
+
+// Fd returns the integer unix file descriptor referencing the open file.
+func (e *Epoll) Fd() int {
+	return e.fd
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err == nil {
+		e.sysEvents = append(e.sysEvents, unix.EpollEvent{})
```

Comment:
Do we need to protect e.sysEvents by a mutex in case multiple go-routines call e.Register() at the same time?

Reply:
I don't believe we can use a per-instance `e.sysEvents` within `Wait()` either without wrapping the wait syscall in the mutex as well (which is presumably not desirable), since multiple syscalls could attempt to write to the same buffer. So I am inclined to do what I mentioned above (https://github.com/snapcore/snapd/pull/12963#discussion_r1264106551), and replace `e.sysEvents` with a `registeredFdCount` variable, which would be protected by a mutex within `Register()` and `Deregister()`.

Reply:
The latest commit should address this, and add some concurrency unit tests.